### ISSUE TITLE
How to install and use

### ISSUE THREAD

**Issue by thb007** [nin9s/elk-hole#1]
I am not an engineer engaged in programming, so I can't read the tutorial you wrote. You can write a one-click installation just like installing pi.hole!

**Comment by nin9s** [comment]
Unfortunately setting up an elk stack is more than the one click install for pihole. If you have managed to set up logstash, kibana and elasticsearch (how to from @skaldenhooven is linked) the installation of elk-hole is pretty straight forward as the guide describes.

Where exactly are you stuck?

**Comment by ChrisRBe** [comment]
Hi @nin9s,

if I manage to find the time on the weekend ... I might take a look at adding a Dockerfile based on the official ELK image, adding your ELK related components to it, so that you can have that part abstracted.

**Comment by nin9s** [comment]
That would be great! I still need to learn a lot about docker so: thank you!

**Comment by ChrisRBe** [comment]
Hi @nin9s 
I created an initial draft here https://github.com/ChrisRBe/elk-hole/pull/1
It's not yet finished as I was not able to test it properly since I'm running pi-hole on my raspbi and was not able to get filebeat running yet.

**Comment by nin9s** [comment]
trying to get some knowledge on howto use a dockerfile with a container etc. 
Am I able to help you with filebeat? running pihole on a native debian vm here.

**Comment by ChrisRBe** [comment]
you could install Docker (apt-get install docker-ce) on that or a different VM, get the changes from my PR (namely the Dockerfile and the small change to the config file. (I could also create the PR already here in your repo.)
Then, run `docker build -t elk-hole .` in the directory with the Dockerfile. 
Once the Image is built, you can run the image like this:
`docker run -p 5601:5601 -p 9200:9200 -p 5044:5044 -it --name elk elk-hole`

This should start you Elasticsearch, Logstash, Kibana with the correct settings. You can connect to Kibana on 127.0.0.1:5601 to load the config.
Depending if you run this on the same VM or not, you have to configure filebeat appropiately (so either also 127.0.0.1:5044, or the remote IP, which should be 10.0.2.2 in case you're running everything locally in VirtualBox (or whatever config you run :)).
If you run VirtualBox locally, and you chose to run a second one, don't forget to set up port forwarding on the Elasticsearch VM. 
I will check how to apply the kibana config via command line, adding this to the Dockerfile as well as a further step. 
Let me know if this makes any sense to you. I will try to get back to you latest on Saturday.

**Comment by nin9s** [comment]
@ChrisRBe sorry didnt have the time so far to test it out ... I'll keep you posted asap

**Comment by ChrisRBe** [comment]
no worries. It looks like I won't be able to automatically configure the Kibana Dashboard as its config is stored in Eleasticsearch, rather than somewhere in Kibana.
Oh well. Will do a once over of the Dockerfile tomorrow I guess and create the PR here.

**Comment by nin9s** [comment]
Can't one just load/import the dashboard via the api after starting the stack?

**Comment by ChrisRBe** [comment]
From what I know, yes. You should be able to import the dashboard via REST API [Kibana Dashboard REST API](https://www.elastic.co/guide/en/kibana/6.5/dashboard-import-api-import.html) but I was not able to get it working yet on my side.

**Comment by nin9s** [comment]
from what I learned it should be no more than:
```
kibana:
   command: curl -XPOST localhost:5601/api/kibana/dashboards/import -H 'kbn-xsrf:true' -H 'Content-type:application/json' -d @elk-hole.json
```

**Comment by ChrisRBe** [comment]
I don't see it working from within the Dockerfile. when creating the Dockerfile you're not actually starting up Kibana ... I mean, you probably could ... 
From this I believe it to be best to load the dashboard somehow when starting up everything.

**Comment by nin9s** [comment]
Can we execute a script and call the api command within?

**Comment by ChrisRBe** [comment]
I found this example from elastic themselves:
https://github.com/elastic/stack-docker

basically .. you would run a second instance that is used to setup Kibana, etc. 

via
https://stackoverflow.com/questions/47024399/kibana-and-docker-compose-automatically-load-in-saved-object

**Comment by nin9s** [comment]
Thanks, I will have a look if I have some spare time left but you can cross
check it as well of course :)

ChrisRBe <notifications@github.com> schrieb am Mo. 22. Apr. 2019 um 12:59:

**Comment by josh-thurston** [comment]
Hi @nin9s,

Stuck on somewhat of a modified version.

I setup Pi-Hole on Ubuntu 18.  
- Filebeat is installed.
- Using your filebeat.yml
- Using 99-pihole-log-facility.conf

I setup ELK stack on Ubuntu 18
- Logstash setup with 20-dns-syslog.conf and dns
- I did not have a patterns directory in /etc/logstash. so i created it and placed dns in that folder.

Kibana setup on same Ubuntu system as elasticsearch and logstash
- Management --> Kibana --> Saved Objects
- imported the following v7.1 files (hope this is correct)
- elk-hole-dash_enhanced.json success
- elk-hole-dash.json success
- elk-hole-vis_enhanced_fix.json didnt work.  Index Pattern Conflicts
- elk-hole-vis.json didnt work. . Index Pattern Conflicts
- in Saved Objects, i did a search for logstash-syslog-dns* and did not have any results

Dev Tools
- Pasted logstash-syslog-dns-index.template_ELK7.x.json
- Is there anything i need to do after paste?  Save? Run? 
- I dont know how to releoad the field index.  I dont see an index listed for this setup
- Dashboards - No data to display on Requests vs piholed.  All the other visuals say 'could not locate that visualization'

Any guidance would be much appreciated.

/JT

**Comment by nin9s** [comment]
Hi @josh-thurston 

> I did not have a patterns directory in /etc/logstash. so i created it and placed dns in that folder.

correct

> imported the following v7.1 files (hope this is correct)

yes, as long as you are using elk 7.1-7.3. forgot to rename the folders, sorry ...

> elk-hole-vis_enhanced_fix.json didnt work. Index Pattern Conflicts
> elk-hole-vis.json didnt work. . Index Pattern Conflicts

can you show me an example screenshot please?

> in Saved Objects, i did a search for logstash-syslog-dns* and did not have any results

you need to search for the index pattern logstash-syslog-dns* in "Kibana -> Index Patterns", not in "Kibana -> Saved Objects"

> Pasted logstash-syslog-dns-index.template_ELK7.x.json
> Is there anything i need to do after paste? Save? Run?

there is green triangle in the upper right corner after pasting the pattern into the console, this actually "runs" the pattern/mapping and puts them into the index (thanks for mentioning this as I really missed to explicitely state this in the howto)

> I dont know how to releoad the field index. I dont see an index listed for this setup

Management -> Index patterns -> logstash-syslog-dns*
if ths pattern was created successfully there should be 79 fields after pressing the refresh arrow in the upper right corner of the page

this requires that the index was successfully created before meaning you need at least one line of log data processed by logstash. I think this one is the actual error in your setup as of now and we need to fix that first.

> Dashboards - No data to display on Requests vs piholed. All the other visuals say 'could not locate that visualization'

this might be due to :
> elk-hole-vis_enhanced_fix.json didnt work. Index Pattern Conflicts
> elk-hole-vis.json didnt work. . Index Pattern Conflicts

**Comment by josh-thurston** [comment]
Thanks for the fast reply @nin9s,

Here is some detail for you.  Sorry if this is overkill

Kibana, Logstash, Elasticsearch, Filebeat are all the most recent version as of downloading yesterday.

Screenshot when importing elk-hole-vis_enhanced_fix.json --> result Pattern Conflicts

![Screen Shot 2019-09-18 at 2 23 16 PM](https://user-images.githubusercontent.com/36908151/65179207-eefe5f00-da1f-11e9-9437-46026c992ac1.png)

No Items found when searching for logstash-syslog-dns* in Index Patterns

![Screen Shot 2019-09-18 at 2 27 16 PM](https://user-images.githubusercontent.com/36908151/65179465-72b84b80-da20-11e9-9fd1-6c2cb23b69af.png)

In Dev Tools, Clicked Green triangle.   Result

![Screen Shot 2019-09-18 at 2 28 32 PM](https://user-images.githubusercontent.com/36908151/65179552-9f6c6300-da20-11e9-8753-a2d4df676a98.png)

After reading through your thorough response.  I think I agree, the issue is with my index pattern conflict.

**Comment by nin9s** [comment]
> After reading through your thorough response. I think I agree, the issue is with my index pattern conflict.

I have to correct my statement somehow as the conflict itself  is not the problem but the conflict sources from the missing index pattern (you got the idea)

Can you please verify and confirm if you actually received any data via logstash? If not, the index wont get created you wont be able to create the pattern/import the visualizations

logstash and elasticsearch running without error? can you paste the input/output section of 20-dns-syslog.conf please

**Comment by josh-thurston** [comment]
hours of researching and troubleshooting, and i am almost there.

Issue was that filebeat was not establishing a connection to logstash.  Cause (i think) was that i had the 20-dns-syslog.conf and also logstash.conf files.  They had different configurations.  Renamed logstash.conf to logstash.conf.old and renamed 20-dns-syslog.conf to logstash.conf.  

Connection was established.
Index Management shows logstash-syslog-dns-2019..09
Vis and Dashboard successfuly import to Saved Objects

Index patterns does not show logstash-syslog-dns* listed when i search?  Can't figure this out
Still dont have all the dashboard working.  Only two work:
1. piholed percent - pihole
2. requests vs piholed - pihole

Any ideas how what direction to troubleshoot next?

I am wondering if it is initializing the index pattern the right way

**Comment by josh-thurston** [comment]
Tried another step and made more progress.

index patterns --> Search for logstash-syslog-dns*
Select Create index patterns
Enter logstash-syslog-dns* in the index pattern box --> Next Step
Choose @timestamp --> Create index pattern

Once the index is created, verify that 79 fields are listed
Click Refresh Field List (curved arrows) top right.

Dashboards are still not showing all the data.  "Could not locate that visualization..."

**Comment by josh-thurston** [comment]
Finally Done!
Import the viz files into saved objects.  It will possibly give an index conflict message.  

![Screen Shot 2019-09-18 at 9 32 45 PM](https://user-images.githubusercontent.com/36908151/65208850-e297f780-da5b-11e9-98f6-c52e3d49f66b.png)

Below the message will be 2 rows.  Select logstash-syslog-dns* from the dropdown.  --> Confirm all changes. 

![Screen Shot 2019-09-18 at 9 33 45 PM](https://user-images.githubusercontent.com/36908151/65208911-13782c80-da5c-11e9-8bce-52df3ea64438.png)

Now revisit / refresh the dashboard and everything should be working!

**Comment by nin9s** [comment]
Thank you josh
I think I have to edit the tutorial and describe the steps in a more detailed fashion.

Your pain simply sourced of the fact that the index pattern wasn’t created before importing the jsons.

There were two problems
1. there was no data 
2. the pattern wasn’t created 

I guess the fact that some vis where not working correctly was due to the fact that the index pattern only gets applied during index creation(!) and not afterwards. So if there are fields with the wrong type you get an error afterwards 

I’ll rethink the order and steps of the tutorial. 
Thanks again for your effort, I will use that for describing the steps a little better

**Comment by nin9s** [comment]
@josh by the way

Your guessed issue with the naming of the conf files shouldn’t be a an issue, at least not by design.
Logstash processes multiple conf files in alphabetical order meaning if you prepend a number in front of the name, they will get procesSed from lower to higher. The only thing you need to make sure than is to verify that a input /output/filter section of these very files do not override a input/output/filter section of a file beeing processed later. I did this via tags and you probably need such a criteria in your original logstash.conf. Otherwise you won’t be able to process different types of data.

**Comment by nin9s** [comment]
@josh-thurston 
I've updated the tutorial. If you have some spare time I would highly appreciate it if you could have a look and tell me if there is something elementary missing or in the wrong order.

**Comment by josh-thurston** [comment]
@nin9s i forked the repo and made updates to the README.md on my fork.  If you are good with those changes we can merge into your master repo.

https://github.com/josh-thurston/elk-hole

I also created a repo with a script to setup an elk stack with some basic config instructions to get it running.  All pre-work that needs to be done before setting up Pi-hole and Elk-hole.
 
Tonight I will create an additional repo with a script to prepare the Raspberry pi to run Filebeat.  Looks like most of the information I am finding on the web is old-ish from 2015, 2016

On another topic.  I am wondering how hard it would be to update elk-hole to pull DHCP logs for people that are running DHCP on their Pi-hole system.

**Comment by nin9s** [comment]
@josh-thurston READNE.md looks great, thanks a lot. To be honest im still learning github and its mechanics. What is the best way to merge your file into my repo?
If I really have much time I can try to find out how we can build this via docker maybe 

> On another topic. I am wondering how hard it would be to update elk-hole to pull DHCP logs for people that are running DHCP on their Pi-hole system.

that wouldnt be too hard as its just another file filebeat can process relatively easily. I just never used the pi as dhcp server but that shouldnt be an issue ;)
Do you want to show dhcp logs simply one per line in a table or something?

**Comment by ratnamg** [comment]
Folks
I feel I am close to my installation after a week of trial and error with the ELK installation. 

When I use the 20-dns-syslog.conf and start logstash. I am getting the following error. Given the error, I am assuming its an error in either the 20-dns-syslog.conf or a file in the patterns . Can I please get some help with this?

[2020-11-30T22:30:34,655][ERROR][logstash.javapipeline    ][main] Pipeline error {:pipeline_id=>"main", :exception=>#<Grok::PatternError: pattern %{DNSMASQPREFIX} not defined>, :backtrace=>["/usr/local/Cellar/logstash-full/7.10.0/libexec/vendor/bundle/jruby/2.5.0/gems/jls-grok-0.11.5/lib/grok-pure.rb:123:in `block in compile'", "org/jruby/RubyKernel.java:1442:in `loop'", "/usr/local/Cellar/logstash-full/7.10.0/libexec/vendor/bundle/jruby/2.5.0/gems/jls-grok-0.11.5/lib/grok-pure.rb:93:in `compile'", "/usr/local/Cellar/logstash-full/7.10.0/libexec/vendor/bundle/jruby/2.5.0/gems/logstash-filter-grok-4.3.0/lib/logstash/filters/grok.rb:288:in `block in register'", "org/jruby/RubyArray.java:1809:in `each'", "/usr/local/Cellar/logstash-full/7.10.0/libexec/vendor/bundle/jruby/2.5.0/gems/logstash-filter-grok-4.3.0/lib/logstash/filters/grok.rb:282:in `block in register'", "org/jruby/RubyHash.java:1415:in `each'", "/usr/local/Cellar/logstash-full/7.10.0/libexec/vendor/bundle/jruby/2.5.0/gems/logstash-filter-grok-4.3.0/lib/logstash/filters/grok.rb:277:in `register'", "org/logstash/config/ir/compiler/AbstractFilterDelegatorExt.java:75:in `register'", "/usr/local/Cellar/logstash-full/7.10.0/libexec/logstash-core/lib/logstash/java_pipeline.rb:228:in `block in register_plugins'", "org/jruby/RubyArray.java:1809:in `each'", "/usr/local/Cellar/logstash-full/7.10.0/libexec/logstash-core/lib/logstash/java_pipeline.rb:227:in `register_plugins'", "/usr/local/Cellar/logstash-full/7.10.0/libexec/logstash-core/lib/logstash/java_pipeline.rb:586:in `maybe_setup_out_plugins'", "/usr/local/Cellar/logstash-full/7.10.0/libexec/logstash-core/lib/logstash/java_pipeline.rb:240:in `start_workers'", "/usr/local/Cellar/logstash-full/7.10.0/libexec/logstash-core/lib/logstash/java_pipeline.rb:185:in `run'", "/usr/local/Cellar/logstash-full/7.10.0/libexec/logstash-core/lib/logstash/java_pipeline.rb:137:in `block in start'"], "pipeline.sources"=>["/usr/local/etc/logstash/conf.d/20-dns-syslog.conf"], :thread=>"#<Thread:0x3ecd722@/usr/local/Cellar/logstash-full/7.10.0/libexec/logstash-core/lib/logstash/pipeline_action/create.rb:54 run>"}
[2020-11-30T22:30:34,660][INFO ][logstash.javapipeline    ][main] Pipeline terminated {"pipeline.id"=>"main"}
[2020-11-30T22:30:34,685][ERROR][logstash.agent           ] Failed to execute action {:id=>:main, :action_type=>LogStash::ConvergeResult::FailedAction, :message=>"Could not execute action: PipelineAction::Create<main>, action_result: false", :backtrace=>nil}
[2020-11-30T22:30:35,057][INFO ][logstash.agent           ] Successfully started Logstash API endpoint {:port=>5141}
[2020-11-30T22:30:36,556][INFO ][logstash.javapipeline    ][.monitoring-logstash] Pipeline terminated {"pipeline.id"=>".monitoring-logstash"}
[2020-11-30T22:30:37,106][INFO ][logstash.runner          ] Logstash shut down.

**Comment by nin9s** [comment]
what is the content of `/etc/logstash/patterns/dns`

**Comment by ratnamg** [comment]
Here is the contents of /etc/logstash/patterns/dns

HOSTNAMEPTR \b(?:[\._0-9A-Za-z][0-9A-Za-z-]{0,62})(?:\.(?:[\._0-9A-Za-z][0-9A-Za-z-]{0,62}))*(\.?|\b)
NODATA NODATA-[[:word:]]+
SRV _+.+\S
FQDN \b(?:[\w-][\w-]{0,62})(?:\.(?:[\w-][\w-]{0,62}))*(\.?|\b)
DNSMASQPREFIX %{SYSLOGTIMESTAMP:date} %{SYSLOGPROG}: %{INT:logrow} %{IP:source_host}\/%{POSINT:source_port}
URI %{URIPROTO}:(//)?(?:%{URIHOST})?(?:%{URIPATH})?

**Comment by ratnamg** [comment]
I figure it out. My installation was under /usr/local/etc.... 

Just changed the 20-dns-syslog.conf file input section.
